File: 백준/Silver/1874. 스택 수열/스택 수열.py
n = int(input())
import sys
stack,result = [], []
## 실패 조건이 무엇일까
## 1 2 5 3 4 -> 1 ,2 정상 5 스택 3 정상 4 스택 -> 
countNum = 1
for i in range(n):
    num = int(sys.stdin.readline())
    ## 스택 쌓기
    while countNum <= num :
        stack.append(countNum)
        result.append('+')
        countNum+=1

    
    ## 스택에서 빼기
    if stack[-1] == num:
        stack.pop()
        result.append('-')
    ## top이 맞지 않은 순간 실패
    else :
        print('NO')
        exit()   
print('\n'.join(result))

File: 백준/Silver/1874. 스택 수열/스택 수열.cc
#include <iostream>
using namespace std;
#include <string>
#include <stack>
string solution() {
	stack <int> st;
	int n; cin >> n;

	int current = 1;
	string result="";
	for (int i = 0; i < n; i++) {
		int num; cin >> num;

		while (current <= num) {
			st.push(current++);
			result += "+\n";
		}
		if (!st.empty() && st.top() == num) {
			st.pop();
			result += "-\n";
		}
		else {
			return "NO";
		}
	}
	return result;
}
int main() {
	ios_base::sync_with_stdio(0);
	cin.tie(0);
	cout.tie(0);

	// push : +, pop: -

	cout << solution();
}
